File: sdk/java/src/main/java/dev/udb/client/UdbStorageClient.java
package dev.udb.client;

import com.udb.core.storage.services.v1.DeleteFileRequest;
import com.udb.core.storage.services.v1.DeleteFileResponse;
import com.udb.core.storage.services.v1.FinalizeUploadRequest;
import com.udb.core.storage.services.v1.FinalizeUploadResponse;
import com.udb.core.storage.services.v1.GetDownloadUrlRequest;
import com.udb.core.storage.services.v1.GetDownloadUrlResponse;
import com.udb.core.storage.services.v1.GetFileRequest;
import com.udb.core.storage.services.v1.GetFileResponse;
import com.udb.core.storage.services.v1.ListFilesRequest;
import com.udb.core.storage.services.v1.ListFilesResponse;
import com.udb.core.storage.services.v1.RegisterUploadRequest;
import com.udb.core.storage.services.v1.RegisterUploadResponse;
import com.udb.core.storage.services.v1.StorageServiceGrpc;
import com.udb.core.storage.services.v1.UpdateFileRequest;
import com.udb.core.storage.services.v1.UpdateFileResponse;
import io.grpc.Channel;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Objects;

/**
 * Blocking facade over the native {@code StorageService}. It rides the shared
 * control-plane channel and attaches the project {@link UdbMetadata} headers to
 * every call. The raw generated stub stays reachable via {@link #stub()}.
 */
public final class UdbStorageClient {
  @FunctionalInterface
  public interface HttpPut {
    void put(String url, byte[] data, String contentType);
  }

  public record UploadFileOptions(
      String contentType,
      String fileType,
      String referenceId,
      String referenceType,
      Boolean isPublic,
      int expiresInMinutes,
      String checksum,
      String etag) {
    public static UploadFileOptions defaults() {
      return new UploadFileOptions(
          "application/octet-stream", "", "", "", null, 0, "", "");
    }

    public UploadFileOptions {
      contentType =
          contentType == null || contentType.isBlank() ? "application/octet-stream" : contentType;
      fileType = fileType == null ? "" : fileType;
      referenceId = referenceId == null ? "" : referenceId;
      referenceType = referenceType == null ? "" : referenceType;
      expiresInMinutes = Math.max(0, expiresInMinutes);
      checksum = checksum == null ? "" : checksum;
      etag = etag == null ? "" : etag;
    }
  }

  private static final HttpClient HTTP = HttpClient.newHttpClient();

  private final StorageServiceGrpc.StorageServiceBlockingStub stub;
  private final UdbMetadataRef metadata;
  private final HttpPut httpPut;

  UdbStorageClient(Channel channel, UdbMetadata metadata) {
    this(channel, metadata, UdbCredentials.fromMetadata(metadata));
  }

  UdbStorageClient(Channel channel, UdbMetadata metadata, UdbCredentials credentials) {
    this(channel, new UdbMetadataRef(metadata), credentials, UdbStorageClient::defaultPut);
  }

  UdbStorageClient(Channel channel, UdbMetadataRef metadata, UdbCredentials credentials) {
    this(channel, metadata, credentials, UdbStorageClient::defaultPut);
  }

  UdbStorageClient(
      Channel channel, UdbMetadataRef metadata, UdbCredentials credentials, HttpPut httpPut) {
    this.metadata = Objects.requireNonNull(metadata, "metadata");
    this.httpPut = Objects.requireNonNull(httpPut, "httpPut");
    this.stub =
        StorageServiceGrpc.newBlockingStub(channel)
            .withInterceptors(UdbClient.credentialInterceptor(metadata, credentials));
  }

  /** The raw generated blocking stub (never hidden). */
  public StorageServiceGrpc.StorageServiceBlockingStub stub() {
    return stub;
  }

  /** Reserve an upload slot and obtain a presigned PUT URL + file id. */
  public RegisterUploadResponse registerUpload(RegisterUploadRequest request) {
    return stub.registerUpload(request);
  }

  public FinalizeUploadResponse uploadFile(String filename, byte[] data) {
    return uploadFile(filename, data, UploadFileOptions.defaults());
  }

  /**
   * Upload bytes through the broker-owned two-phase storage workflow:
   * RegisterUpload -> HTTP PUT to the presigned URL -> FinalizeUpload.
   *
   * <p>No proof Get/List is hidden here; an empty upload URL fails closed before
   * the PUT/finalize steps.
   */
  public FinalizeUploadResponse uploadFile(
      String filename, byte[] data, UploadFileOptions options) {
    Objects.requireNonNull(filename, "filename");
    Objects.requireNonNull(data, "data");
    UdbMetadata current = metadata.current();
    UploadFileOptions opts = options == null ? UploadFileOptions.defaults() : options;
    RegisterUploadRequest.Builder register =
        RegisterUploadRequest.newBuilder()
            .setTenantId(current.tenantId())
            .setProjectId(current.projectId())
            .setFilename(filename)
            .setContentType(opts.contentType())
            .setFileType(opts.fileType())
            .setReferenceId(opts.referenceId())
            .setReferenceType(opts.referenceType())
            .setExpiresInMinutes(opts.expiresInMinutes())
            .setSizeBytes(data.length);
    if (opts.isPublic() != null) {
      register.setIsPublic(opts.isPublic());
    }

    RegisterUploadResponse registered = registerUpload(register.build());
    if (registered.getUploadUrl().isBlank()) {
      throw new IllegalStateException("RegisterUpload returned no upload_url");
    }
    httpPut.put(registered.getUploadUrl(), data, opts.contentType());

    FinalizeUploadRequest.Builder finalize =
        FinalizeUploadRequest.newBuilder()
            .setTenantId(current.tenantId())
            .setFileId(registered.getFileId())
            .setContentType(opts.contentType())
            .setFileType(opts.fileType())
            .setReferenceId(opts.referenceId())
            .setReferenceType(opts.referenceType())
            .setSizeBytes(data.length);
    if (opts.isPublic() != null) {
      finalize.setIsPublic(opts.isPublic());
    }
    if (!opts.checksum().isBlank()) {
      finalize.setChecksum(opts.checksum());
    }
    if (!opts.etag().isBlank()) {
      finalize.setEtag(opts.etag());
    }
    return finalizeUpload(finalize.build());
  }

  /** Confirm an upload completed; transitions the file out of PENDING. */
  public FinalizeUploadResponse finalizeUpload(FinalizeUploadRequest request) {
    return stub.finalizeUpload(request);
  }

  /** Mint a presigned GET URL for downloading a stored object. */
  public GetDownloadUrlResponse getDownloadUrl(GetDownloadUrlRequest request) {
    return stub.getDownloadUrl(request);
  }

  /** Fetch a file's metadata by id. */
  public GetFileResponse getFile(GetFileRequest request) {
    return stub.getFile(request);
  }

  /** Update mutable file metadata (name, tags, etc.). */
  public UpdateFileResponse updateFile(UpdateFileRequest request) {
    return stub.updateFile(request);
  }

  /** Delete a file (and schedule object GC). */
  public DeleteFileResponse deleteFile(DeleteFileRequest request) {
    return stub.deleteFile(request);
  }

  /** List files for the tenant, with paging/filter on the request. */
  public ListFilesResponse listFiles(ListFilesRequest request) {
    return stub.listFiles(request);
  }

  private static void defaultPut(String url, byte[] data, String contentType) {
    HttpRequest request =
        HttpRequest.newBuilder(URI.create(url))
            .header("content-type", contentType)
            .PUT(HttpRequest.BodyPublishers.ofByteArray(data))
            .build();
    try {
      HttpResponse<Void> response = HTTP.send(request, HttpResponse.BodyHandlers.discarding());
      if (response.statusCode() < 200 || response.statusCode() >= 300) {
        throw new IllegalStateException("upload PUT failed with HTTP " + response.statusCode());
      }
    } catch (IOException err) {
      throw new UncheckedIOException(err);
    } catch (InterruptedException err) {
      Thread.currentThread().interrupt();
      throw new IllegalStateException("upload PUT interrupted", err);
    }
  }
}

File: sdk/csharp/Udb.Client/UdbStorageClient.cs
using Grpc.Core;
using StorageV1 = Udb.Core.Storage.Services.V1;

namespace Udb.Client;

/// <summary>
/// Ergonomic async facade over the native <c>StorageService</c>. Every wrapper
/// applies the shared <see cref="UdbProject"/> metadata headers and returns the
/// unwrapped response task. The raw generated client stays reachable via
/// <see cref="Raw"/> for RPCs / overloads not wrapped here.
/// </summary>
public sealed class UdbStorageClient
{
    public sealed record UploadFileOptions(
        string ContentType = "application/octet-stream",
        string FileType = "",
        string ReferenceId = "",
        string ReferenceType = "",
        bool? IsPublic = null,
        int ExpiresInMinutes = 0,
        string Checksum = "",
        string Etag = "");

    private static readonly HttpClient DefaultHttpClient = new();

    private readonly StorageV1.StorageService.StorageServiceClient _client;
    private readonly Func<Metadata> _headers;
    private readonly Func<string, byte[], string, CancellationToken, Task> _putBytes;

    internal UdbStorageClient(
        StorageV1.StorageService.StorageServiceClient client,
        Func<Metadata> headers,
        Func<string, byte[], string, CancellationToken, Task>? putBytes = null)
    {
        _client = client;
        _headers = headers;
        _putBytes = putBytes ?? PutBytesAsync;
    }

    /// <summary>Raw generated storage service client.</summary>
    public StorageV1.StorageService.StorageServiceClient Raw => _client;

    /// <summary>Register a new upload and obtain a pre-signed upload URL.</summary>
    public Task<StorageV1.RegisterUploadResponse> RegisterUploadAsync(
        StorageV1.RegisterUploadRequest request, CancellationToken ct = default)
        => _client.RegisterUploadAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Finalize an upload after the object has been written to the store.</summary>
    public Task<StorageV1.FinalizeUploadResponse> FinalizeUploadAsync(
        StorageV1.FinalizeUploadRequest request, CancellationToken ct = default)
        => _client.FinalizeUploadAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>
    /// Upload bytes through the storage two-phase workflow:
    /// RegisterUpload -> HTTP PUT to the presigned URL -> FinalizeUpload. No
    /// hidden Get/List proof calls are issued.
    /// </summary>
    public async Task<StorageV1.FinalizeUploadResponse> UploadFileAsync(
        string filename,
        byte[] data,
        UploadFileOptions? options = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(filename);
        ArgumentNullException.ThrowIfNull(data);
        var opts = options ?? new UploadFileOptions();
        var contentType = string.IsNullOrWhiteSpace(opts.ContentType)
            ? "application/octet-stream"
            : opts.ContentType;
        var headers = _headers();
        var register = new StorageV1.RegisterUploadRequest
        {
            TenantId = Header(headers, "x-tenant-id"),
            ProjectId = Header(headers, "x-udb-project-id"),
            Filename = filename,
            ContentType = contentType,
            FileType = opts.FileType,
            ReferenceId = opts.ReferenceId,
            ReferenceType = opts.ReferenceType,
            ExpiresInMinutes = Math.Max(0, opts.ExpiresInMinutes),
            SizeBytes = data.LongLength,
        };
        if (opts.IsPublic.HasValue)
        {
            register.IsPublic = opts.IsPublic.Value;
        }

        var registered = await _client.RegisterUploadAsync(register, headers, cancellationToken: ct).ResponseAsync
            .ConfigureAwait(false);
        if (string.IsNullOrWhiteSpace(registered.UploadUrl))
        {
            throw new InvalidOperationException("RegisterUpload returned no upload_url");
        }

        await _putBytes(registered.UploadUrl, data, contentType, ct).ConfigureAwait(false);

        var finalize = new StorageV1.FinalizeUploadRequest
        {
            TenantId = Header(headers, "x-tenant-id"),
            FileId = registered.FileId,
            ContentType = contentType,
            FileType = opts.FileType,
            ReferenceId = opts.ReferenceId,
            ReferenceType = opts.ReferenceType,
            SizeBytes = data.LongLength,
        };
        if (opts.IsPublic.HasValue)
        {
            finalize.IsPublic = opts.IsPublic.Value;
        }
        if (!string.IsNullOrEmpty(opts.Checksum))
        {
            finalize.Checksum = opts.Checksum;
        }
        if (!string.IsNullOrEmpty(opts.Etag))
        {
            finalize.Etag = opts.Etag;
        }

        return await _client.FinalizeUploadAsync(finalize, headers, cancellationToken: ct).ResponseAsync
            .ConfigureAwait(false);
    }

    /// <summary>Get a pre-signed download URL for a file.</summary>
    public Task<StorageV1.GetDownloadUrlResponse> GetDownloadUrlAsync(
        StorageV1.GetDownloadUrlRequest request, CancellationToken ct = default)
        => _client.GetDownloadUrlAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Get file metadata.</summary>
    public Task<StorageV1.GetFileResponse> GetFileAsync(
        StorageV1.GetFileRequest request, CancellationToken ct = default)
        => _client.GetFileAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Update file metadata.</summary>
    public Task<StorageV1.UpdateFileResponse> UpdateFileAsync(
        StorageV1.UpdateFileRequest request, CancellationToken ct = default)
        => _client.UpdateFileAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>Delete a file (soft delete).</summary>
    public Task<StorageV1.DeleteFileResponse> DeleteFileAsync(
        StorageV1.DeleteFileRequest request, CancellationToken ct = default)
        => _client.DeleteFileAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    /// <summary>List files.</summary>
    public Task<StorageV1.ListFilesResponse> ListFilesAsync(
        StorageV1.ListFilesRequest request, CancellationToken ct = default)
        => _client.ListFilesAsync(request, _headers(), cancellationToken: ct).ResponseAsync;

    private static string Header(Metadata headers, string name)
        => headers.GetValue(name) ?? string.Empty;

    private static async Task PutBytesAsync(
        string url,
        byte[] data,
        string contentType,
        CancellationToken ct)
    {
        using var content = new ByteArrayContent(data);
        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue(contentType);
        using var response = await DefaultHttpClient.PutAsync(url, content, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
    }
}
